server/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -500,15 +503,21 @@ public void updateShardState(final ShardRouting newRouting,
                                  * numbers. To ensure that this is not the case, we restore the state of the local checkpoint tracker by
                                  * replaying the translog and marking any operations there are completed.
                                  */
-                                final Engine engine = getEngine();
-                                engine.restoreLocalCheckpointFromTranslog();
+                                if (seqNoStats().getLocalCheckpoint() < maxSeqNoOfResettingEngine) {
+                                    // The engine was reset before but hasn't restored all existing operations yet.
+                                    // We need to reset the engine again with all the local history.
+                                    getEngine().flush();
+                                    resetEngineUpToSeqNo(Long.MAX_VALUE);
```

Comment:
I don't think we should use `resetEngineUpToSeqNo` here with all its machinery of `SearchOnlyEngine` etc. which only makes sense if there are conflicts in the history. As a primary, we dictate the history here, and can therefore "just" recover from translog (+ refresh afterwards)?